### ISSUE TITLE
The letter I - И or Й?

### ISSUE THREAD

**Issue by TheHockeyist** [TheHockeyist/russian-untransliterator#10]
The letter и is always transliterated as i, and the letter i always represents и. There is no doubt about that. It works fluently in both directions. **И = I. Done.**

**Except, sometimes (especially in older or traditional transliterations), the letter й is also transliterated as i.** There are alternatives, like y or j used to avoid confusion (j is unambiguously й), but if the user chooses i for й, it causes interesting problems.

For example, moi can be мои or мой. Which one is it? -ii can either be ий (as is usually the case) or ии in some words, like России (a grammatical form of Россия). Which one is it? There's no easy way to tell, so we need to resort to asking the user.

After vowels, i is almost certainly й, except in a few cases (especially when grammar comes into play). When in doubt, ask the user whether they meant и or й.

**If searching the text and a j comes up in the Latin text, this is strong evidence that this stage may not be necessary, and that й (and likewise я, е, ё, and ю) was transliterated with j, and that moi really _IS_ мои, for if it was мой, it would have been transliterated as moj.**

Some people transliterate й as y (especially English speakers), which unfortunately confuses it with ы. Thankfully, context can quickly determine which letter was meant for y, as vowel + ы is incredibly unlikely and й is almost certain (this is already implemented in the code). Also, consonant + й is also incredibly unlikely, and so ы is almost certainly meant after consonants.

But й = i is its own breed. I transliterate й as i out of habit, and also partially because that's how I actually learned it in the alphabet. **My name has й, and it was transliterated as i, so it's also a personal thing.**

So, any ideas on addressing this issue?

**Comment by Tymewalk** [comment]
Would "j" ever be used in any case other than й? If not, then all we have to do is check for "j" to see what "i" would end up being.

**Comment by TheHockeyist** [comment]
True. Although j is sometimes used in я, е, ё, and ю, you can never have йа, йе, йо, or йу. Those are automatically corrected, and the few words with йе and йо are corrected back. Thus, j = й, and the other cases are automatically corrected.

If a j appears in transliterated text, it's a sign that the user uses i for и only.

**Comment by Tymewalk** [comment]
@TheHockeyist Got it.

I recommend we use `re` so we can just do this:

```
if re.search("j", latin_string):
    uses_j = True
else:
    uses_j = False
```

And later we check if `uses_j` was set to `True` or not when deciding how to transliterate "i".

**Comment by TheHockeyist** [comment]
Okay. Add it to the code then!

**Comment by TheHockeyist** [comment]
Well done, Ty.

Then if there are no j's anywhere, we just ask the user.

**Comment by TheHockeyist** [comment]
We now bring all further discussion to #6 (Asking for clarification.) and we're done.

**Comment by TheHockeyist** [comment]
Reopened - still relevant. I had text saved here but then my computer decided to go to another tab somehow, and so welp, must retype.

It wouldn't make sense to convert all the other Latin letters and then determine what i is. It's ambiguous before and after. Leaving it last isn't doing us any favors. I personally like to convert everything to Cyrillic first and then correct the errors.

Here are the rules I propose:

**Rule 1: Initially convert all i's in the text to и.**

**Rule 2: Check for j's**

**Rule 2a: If there are j's, initially convert all j's in the text to й. Rule 2b doesn't need to kick in.**

**Rule 2b: If there are no j's, that means i can stand for either и or й. The trick now is to figure out which и's need to switch to й's in the following rules 3-7.**

**Rule 3: The letter й almost always occurs in diphthongs.** This is a rule. The only exceptions are those loanwords like йогурт and Йемен. So if an i is after a consonant, it's an и. End of story. The letter и is far more common than й is for this reason. **If an i is after a vowel, expect it to be й. But don't follow this rule blindly.** The rules below help clarify things and give situations where it can be и.

**Rule 4: After an ы or е, it is almost certainly й.** Exceptions are like loanwords. At the end of a word, they're almost guaranteed to be й. I can tell you that. You never see и at the end of a word in these circumstances. It just looks too strange.

**Rule 5: After an а or о, it is likely to be й, but less so.** You can still trust that it's likely to be й, but there is a higher risk for mistake. Common exceptions include words like аист or мои (мой is also a word, so we're in more trouble here - just ask the user in this case.).

**Rule 6: у and ю aren't very common.** The most common example I can think of are imperatives like жуй, etc. and the very hard to pronounce дюйм. You just don't see these ones too often. ё is also seldom seen, and I don't have enough data on it to give you good conclusions. You'll never get it after э unless in very rare loanwords. I've never seen эй, but it is expected more than эи in my opinion.

**Rule 7: и is a hard case to get right.** For adjectives, -й is correct, like хороший. For nouns, -и is more likely to be correct, like России. There are names like Дмитрий, though. In these cases, you can assume й initially, but check with the user later to make sure that it's correct and you don't run into a case like армии or России.

**Comment by TheHockeyist** [comment]
Actively working on this issue right now.

**Comment by Tymewalk** [comment]
I'd say we can safely replace consonant-i with "и" and ыi/ei with "й".

We'd have to check the others.

**Comment by TheHockeyist** [comment]
I'd say that is the best way to go.

Also see #6 (Asking for clarification.)
